filters/include/pcl/filters/experimental/cartesian_filter.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/filters/experimental/transform_filter.h>
+
+#include <limits>
+
+namespace pcl {
+namespace experimental {
+
+/**
+ * \brief CartesianFilter represents the base class for grid filters.
+ * \ingroup filters
+ */
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class CartesianFilter : public TransformFilter<GridStruct, PointT> {
+protected:
+  using PointCloud = typename TransformFilter<GridStruct, PointT>::PointCloud;
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+
+public:
+  /** \brief Empty constructor. */
+  CartesianFilter() {}
+
+  /** \brief Destructor. */
+  ~CartesianFilter() {}
+
+  /** \brief Set to true if all fields need to be downsampled, or false if just XYZ.
+   * \param[in] downsample the new value (true/false)
+   */
+  inline void
+  setDownsampleAllData(const bool downsample)
+  {
+    downsample_all_data_ = downsample;
+  }
+
+  /** \brief Get the state of the internal downsampling parameter (true if all fields
+   * need to be downsampled, false if just XYZ).
+   */
+  inline bool
+  getDownsampleAllData() const
+  {
+    return downsample_all_data_;
+  }
+
+  /** \brief Set the minimum number of points required for a voxel to be used.
+   * \param[in] min_points_per_voxel the minimum number of points for required for a
+   * voxel to be used
+   */
+  inline void
+  setMinimumPointsNumberPerVoxel(const std::size_t min_points_per_voxel)
+  {
+    min_points_per_voxel_ = min_points_per_voxel;
+  }
+
+  /** \brief Return the minimum number of points required for a voxel to be used.
+   */
+  inline std::size_t
+  getMinimumPointsNumberPerVoxel() const
+  {
+    return min_points_per_voxel_;
+  }
+
+  /** \brief Hash a point to the index of a 3D grid cell.
+   * \param[in] pt the point to hash
+   * \param[in] inverse_leaf_size the array of 1/leaf_size
+   * \param[in] min_b the minimum bin coordinate
+   * \param[in] divb_mul_y division multiplier of the second dimension
+   * \param[in] divb_mul_z division multiplier of the third dimension
+   */
+  inline std::size_t
+  hashPoint(const PointT& pt,
```

Comment:
Feels like the has and overflow related functions could be:
* free-functions
* static functions if required
* member functions if they take data from the class, instead of arguments

Reply:
Done, move to free functions